resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -727,6 +729,10 @@ private[spark] class ApplicationMaster(
             mainMethod.invoke(null, userArgs.toArray)
             finish(FinalApplicationStatus.SUCCEEDED, ApplicationMaster.EXIT_SUCCESS)
             logDebug("Done running user class")
+            if (!unregistered) {
+              unregister(finalStatus, finalMsg)
+              cleanupStagingDir(defaultStagingDir)
+            }
```

Comment:
One question I have is the following comment:
https://github.com/apache/spark/blob/13dddee9a8490ead00ff00bd741db4a170dfd759/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala#L236-L238

This hook is responsible calling `unregister`. So, if the user code didn't explicitly stop the `SparkContext`, this change calls `unregister` before `SparkContext` is stopped. Will this be OK? (from my testing, I didn't find anything wrong so far, but I could be missing something.)